wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,327 @@
+			let target = GetSpellTargetUnit()
+			let effect = target.addEffect(Abilities.voodooAuraTarget, "origin")
+			ForDamageCallback cb = () d ->
+				if target == d.target
+					d.amount -= d.initial * .5
+			addDamageModifier(cb)
+			doAfter(BUFF_DURATION) ->
+				removeDamageModifier(cb)
+				effect.destr()
+
+*/
+package DamageModify
+import AbilityObjEditing
+import DamageDetection
+import OnUnitEnterLeave
+import ObjectIdGenerator
+import LinkedList
+import ClosureTimers
+
+@configurable public constant DAMAGE_ID_NONTRIGGERED = -1
+
+public interface ForDamageCallback
+	function run(Damage d)
+
+let damageModifiers = new LinkedList<ForDamageCallback>
+public function addDamageModifier(ForDamageCallback cb)
+	damageModifiers.add(cb)
+public function removeDamageModifier(ForDamageCallback cb)
+	damageModifiers.remove(cb)
+
+let damageListeners = new LinkedList<ForDamageCallback>
+public function addDamageListener(ForDamageCallback cb)
+	damageListeners.add(cb)
+public function removeDamageListener(ForDamageCallback cb)
+	damageListeners.remove(cb)
+
+/** Deals triggered damage from unit to target.
+	You can specify id if there's any unique id interactions.
+	Damage instance is saved after being destroyed. */
+public function unit.returnDamage(unit target, real amount, bool spell, int id) returns Damage
+	return new Damage(this, target, amount, spell, id)..saveInstance()..modify()..deal()
+
+/** Deals triggered damage from unit to target.
+	You can specify id if there's any unique id interactions. */
+public function unit.damage(unit target, real amount, bool spell, int id)
+	new Damage(this, target, amount, spell, id)..modify()..deal()
+
+/** Deals triggered damage from unit to target. */
+public function unit.damage(unit target, real amount, bool spell)
+	new Damage(this, target, amount, spell, 0)..modify()..deal()
+/** Deals triggered spell damage from unit to target. */
+public function unit.damage(unit target, real amount)
+	new Damage(this, target, amount, true, 0)..modify()..deal()
+
+// ConvertAttackType(7) is a hidden attack type with some unique properties. For
+// more information see goo.gl/9k8tn .
+constant ATTACK_TYPE_UNIVERSAL = ConvertAttackType(7)
+constant DAMAGE_TYPE_CHECK_ID = compiletime(ABIL_ID_GEN.next())
+constant CHEAT_DEATH_ID = compiletime(ABIL_ID_GEN.next())
+
+public class Damage
+	integer id
+	unit source
+	unit target
+	bool spell
+	real initial
+	real amount
+
+	/** check this if you want to save target from damage in all cases */
+	protected bool doSaveTarget = false
+	function saveTarget()
+		doSaveTarget = true
+
+	/** Damage instances are automatically destroyed.
+		Use this to protect them if you need that instances somehow. */
+	function saveInstance()
+		wantSave += 1
+
+	/** Releases Damage instance so it can be destroyed. */
+	function releaseInstance()
+		wantSave -= 1
+		if fired and wantSave <= 0
+			destroy this
+
+	protected int wantSave = 0
+	
+	/** stores how much health were added */
```

Comment:
I don't think this comment is clear. Do you mean "stores the sum of the damage modifiers, after base damage"?